src/main/3.0_3.1/scala/com/crealytics/spark/v2/excel/ExcelTable.scala

```diff
@@ -71,36 +71,49 @@ case class ExcelTable(
       sample = if (sample < 1) 1 else sample
       inputPaths.take(sample).map(_.getPath.toUri)
     }
-    var rows = excelHelper.getRows(conf, paths.head)
-
-    if (rows.isEmpty) { /* If the first file is empty, not checking further */
-      StructType(Seq.empty)
-    } else {
-      /* Prepare field names */
-      val colNames =
-        if (options.header) { /* Get column name from the first row */
-          val r = excelHelper.getColumnNames(rows.next)
-          rows = rows.drop(options.ignoreAfterHeader)
-          r
-        } else { /* Peek first row, then return back */
-          val headerRow = rows.next
-          val r = excelHelper.getColumnNames(headerRow)
-          rows = Iterator(headerRow) ++ rows
-          r
+    var sheetData = excelHelper.getSheetData(conf, paths.head)
+
+    try {
+      if (sheetData.rowIterator.isEmpty) {
+        /* If the first file is empty, not checking further */
+        StructType(Seq.empty)
+      } else {
+        /* Prepare field names */
+        val colNames =
+          if (options.header) {
+            /* Get column name from the first row */
+            val r = excelHelper.getColumnNames(sheetData.rowIterator.next)
+            sheetData = SheetData(sheetData.rowIterator.drop(options.ignoreAfterHeader), sheetData.resourcesToClose)
+            r
+          } else {
+            /* Peek first row, then return back */
+            val headerRow = sheetData.rowIterator.next
+            val r = excelHelper.getColumnNames(headerRow)
+            sheetData = SheetData(Iterator(headerRow) ++ sheetData.rowIterator, sheetData.resourcesToClose)
+            r
+          }
+
+        /* Other files also be utilized (lazily) for field types, reuse field name
+           from the first file */
+        val numberOfRowToIgnore = if (options.header) (options.ignoreAfterHeader + 1) else 0
+        paths.tail.foreach(path => {
+          val newRows = excelHelper.getSheetData(conf, path)
+          sheetData = SheetData(
```

Comment:
There's a lot of copied code here. I wonder at what point we DRY this and put the code that is shared for the different Spark versions into shared code.